In [14]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from scipy import stats
import seaborn as sns
import statsmodels.api as sm
from sklearn import linear_model

### Linear Regression, 1 variable

In [8]:
# Load the data into a pandas dataframe
iris = sns.load_dataset("iris")
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [11]:
X = iris[["petal_length"]] # predictor
y = iris[["petal_width"]] # response

In [12]:
# Linear regression
# Note the swap of X and y
model = sm.OLS(y, X)
results = model.fit()
# Statsmodels gives R-like statistical output
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:            petal_width   R-squared:                       0.967
Model:                            OLS   Adj. R-squared:                  0.967
Method:                 Least Squares   F-statistic:                     4417.
Date:                Fri, 01 Feb 2019   Prob (F-statistic):          1.22e-112
Time:                        12:06:02   Log-Likelihood:                -8.7179
No. Observations:                 150   AIC:                             19.44
Df Residuals:                     149   BIC:                             22.45
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
petal_length     0.3365      0.005     66.463   

**R-squared:** This model explains 96.7% of the variation in petal width.

If **Prob** < 0.05, then is significant.

In [15]:
# coef is slope
# where is my y-intercept?
X = iris["petal_length"]
X = np.vander(X, 2) # add a constant row for the intercept
y = iris["petal_width"]

model = sm.OLS(y, X)
results = model.fit()
print(results.summary())
# x1 = intercept value
# const = slope value
# P = 0; because < 0.05, and therefore statistically significant.

                            OLS Regression Results                            
Dep. Variable:            petal_width   R-squared:                       0.927
Model:                            OLS   Adj. R-squared:                  0.927
Method:                 Least Squares   F-statistic:                     1882.
Date:                Fri, 01 Feb 2019   Prob (F-statistic):           4.68e-86
Time:                        12:13:34   Log-Likelihood:                 24.796
No. Observations:                 150   AIC:                            -45.59
Df Residuals:                     148   BIC:                            -39.57
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1             0.4158      0.010     43.387      0.0

### Formalize relationship between petal width and petal length
`petal_width = 0.41-0.36*(petal_length)`

Linear regression just produces the equation of a straight line, which connects my response variable with the X (predictor).

### Multiple regression

In [17]:
X = iris[["petal_length", "sepal_length"]] # predictors
y = iris[["petal_width"]] # response

In [18]:
# Multiple linear regression
X = iris[["petal_length", "sepal_length"]]
X = sm.add_constant(X) # another way to add a constant row for an intercept
y = iris[["petal_width"]]

model = sm.OLS(y, X)
results = model.fit()
print(results.summary())


                            OLS Regression Results                            
Dep. Variable:            petal_width   R-squared:                       0.929
Model:                            OLS   Adj. R-squared:                  0.928
Method:                 Least Squares   F-statistic:                     962.1
Date:                Fri, 01 Feb 2019   Prob (F-statistic):           3.60e-85
Time:                        12:25:42   Log-Likelihood:                 26.792
No. Observations:                 150   AIC:                            -47.58
Df Residuals:                     147   BIC:                            -38.55
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -0.0090      0.182     -0.049   

When we have more than one predictor, we report Adj. R-squared.

**Adj. R-squared:** This model explains 92.8% of the variation in petal width.

**const** = constant of the intercept.
Slopes of the values of petal length and sepal length.

Since P > 0.05, the intercept is not significant.  Will not be used in equation.

The slope coefficients are statistically significant, because they are both less than 0.05.


### Categorical variables

In [21]:
# dummy-variable regression
dummies = pd.get_dummies(iris["species"])
# Add to the original dataframe
iris = pd.concat([iris, dummies], axis=1) # assign numerical values to the different species
iris.head()

sepal_length  sepal_width  petal_length  petal_width species  setosa  \
0           5.1          3.5           1.4          0.2  setosa       1   
1           4.9          3.0           1.4          0.2  setosa       1   
2           4.7          3.2           1.3          0.2  setosa       1   
3           4.6          3.1           1.5          0.2  setosa       1   
4           5.0          3.6           1.4          0.2  setosa       1   

   versicolor  virginica  setosa  versicolor  virginica  
0           0          0       1           0          0  
1           0          0       1           0          0  
2           0          0       1           0          0  
3           0          0       1           0          0  
4           0          0       1           0          0

In [23]:
X = iris[["petal_length", "sepal_length", "setosa", "versicolor", "virginica"]]
X = sm.add_constant(X) # add row for intercept
y = iris[["petal_width"]]

model = sm.OLS(y, X)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:            petal_width   R-squared:                       0.946
Model:                            OLS   Adj. R-squared:                  0.944
Method:                 Least Squares   F-statistic:                     629.8
Date:                Fri, 01 Feb 2019   Prob (F-statistic):           1.54e-90
Time:                        12:38:37   Log-Likelihood:                 46.705
No. Observations:                 150   AIC:                            -83.41
Df Residuals:                     145   BIC:                            -68.36
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            0.2026      0.102      1.993   

You would be inclined to choose the model that had the lower AIC or BIC value

In [25]:
# Fit the linear model using sklearn
# from sklearn import linear_model
model = linear_model.LinearRegression()
results = model.fit(X, y)

# Print the coefficients
print(results.intercept_, results.coef_)

[0.33766832] [[ 0.          0.23192122 -0.00169337 -0.21113007 -0.21113007  0.00519957
   0.00519957  0.2059305   0.2059305 ]]
